controller/src/test/java/io/pravega/controller/server/rpc/auth/ControllerGrpcAuthFocusedTest.java

```diff
@@ -0,0 +1,478 @@
+import io.pravega.controller.stream.api.grpc.v1.Controller.StreamInfo;
+import io.pravega.controller.stream.api.grpc.v1.Controller.CreateScopeStatus;
+import io.pravega.controller.stream.api.grpc.v1.Controller.TxnId;
+import io.pravega.controller.stream.api.grpc.v1.ControllerServiceGrpc;
+import io.pravega.controller.stream.api.grpc.v1.ControllerServiceGrpc.ControllerServiceBlockingStub;
+import io.pravega.controller.task.Stream.StreamMetadataTasks;
+import io.pravega.controller.task.Stream.StreamTransactionMetadataTasks;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.Timeout;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.net.URI;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+import java.util.Collections;
+import java.util.UUID;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static io.pravega.controller.auth.AuthFileUtils.credentialsAndAclAsString;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * The tests in this class are intended to perform verification of authorization logic in ControllerServiceImpl (the
+ * service implementation of the Controller gRPC interface).
+ */
+public class ControllerGrpcAuthFocusedTest {
+
+    /**
+     * These two members are shared across the tests in this class for efficiency reasons.
+     */
+    private final static ScheduledExecutorService EXECUTOR =
+            ExecutorServiceHelpers.newScheduledThreadPool(20, MethodHandles.lookup().lookupClass() + "-pool");
+    private final static File AUTH_FILE = createAuthFile();
+
+    private final static String DEFAULT_PASSWORD = "1111_aaaa";
+
+    /**
+     * This rule makes sure that the tests in this class run in 10 seconds or less.
+     */
+    @Rule
+    public final Timeout globalTimeout = new Timeout(10, TimeUnit.SECONDS);
+
+    /**
+     * This rule is used later to expect both the exception class and the message.
+     */
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
+
+    private StreamMetadataTasks streamMetadataTasks;
+    private StreamTransactionMetadataTasks streamTransactionMetadataTasks;
+    private Server grpcServer;
+    private ManagedChannel inProcessChannel;
+
+    @AfterClass
+    public static void classTearDown() {
+        if (AUTH_FILE != null && AUTH_FILE.exists()) {
+            AUTH_FILE.delete();
+        }
+        ExecutorServiceHelpers.shutdown(EXECUTOR);
+    }
+
+    @Before
+    public void setup() throws IOException {
```

Comment:
The tests in this class are really short (which is a plus), but this setup is really long. I'm wondering whether we can shorten the per-test setup by making some of these elements common.

Reply:
The tests were pretty fast, as we are:
* creating in-memory Pravega objects (like host store, stream store, etc.), 
* using some mock objects (such as Cluster), and 
* using a special in-process gRPC server, which is used in some of the unit test samples in gRPC repo (for example, https://github.com/grpc/grpc-java/blob/master/examples/src/test/java/io/grpc/examples/header/HeaderServerInterceptorTest.java). 

Running test initializers (@Before method) before each test makes the tests independent and parallelizable. I think the efficiency of the initializers isn't much of a concern in the tests in this class.